pyresample/bucket/__init__.py

```diff
@@ -209,27 +206,39 @@ def _call_pandas_groupby_statistics(self, scipy_method, data, fill_value=None, s
         # Calculate the min of the data falling to each bin
         out_size = self.target_area.size
 
-        # merge into one Dataframe
-        df = dd.concat([dd.from_dask_array(self.idxs), dd.from_dask_array(weights)],
-                       axis=1)
-        df.columns = ['x', 'values']
-
-        if scipy_method == 'min':
-            statistics = df.map_partitions(lambda part: part.groupby(
-                                           np.digitize(part.x,
-                                                       bins=np.linspace(0, out_size, out_size)
-                                                       )
-                                           )['values'].min())
-
-        elif scipy_method == 'max':
-            statistics = df.map_partitions(lambda part: part.groupby(
-                                           np.digitize(part.x,
-                                                       bins=np.linspace(0, out_size, out_size)
-                                                       )
-                                           )['values'].max())
-
-        # fill missed index
-        statistics = (statistics + pd.Series(np.zeros(out_size))).fillna(0)
+        # create the output bins
+        bins = da.linspace(0, out_size-1, out_size).astype('int')
+
+        # sort idxs and weights based on weights
+        if statistic_method == 'min':
+            order = da.map_blocks(np.argsort, weights)
+        elif statistic_method == 'max':
+            order = da.map_blocks(np.argsort, weights)[::-1]
+
+        weights_sorted = da.empty((len(weights) + 1), dtype=weights.dtype)
+        weights_sorted[:-1] = weights[order]
+        weights_sorted[-1] = np.nan
+        idxs_sorted = self.idxs[order]
+
+        extended_bins = da.empty(len(bins) + 1, dtype=bins.dtype)
+        extended_bins[:-1] = bins
+        extended_bins[-1] = np.iinfo(bins.dtype).max  # last bin goes to infinity
+
+        # preallocate buffers to avoid array creation in main loop
+        chunk_size = max(len(weights)//int(1e3), len(bins))
+        mask_buffer = da.empty((chunk_size, len(self.idxs) + 1), dtype=bool)
+        mask_buffer[:, -1] = True
+        statistics = da.empty_like(bins, dtype=np.float64)
+
+        # calculate the statistics in each chunk
+        for low in range(0, len(bins), chunk_size):
```

Comment:
@pnuu  @djhoese Now, I switched to the new method. But, have now idea of applying dask parallel or `da.from_delayed()` to this loop. Another problem is that I don't know the suitable `chunk_size` for the loop, and it is set manually ...